README.md
[![ql.io](http://ql.io/images/ql.io-large.png)](http://ql.io)

ql.io is a declarative, data-retrieval and aggregation gateway for quickly consuming HTTP APIs. See
[ql.io](http://ql.io) for docs, demos and examples. 

![Travis status](https://secure.travis-ci.org/ql-io/ql.io.png)

## How to Build ql.io 

To build ql.io on your own, you need node (version 0.8.0 or later) and npm. In Ubuntu, you need
libexpat-dev additionally. Once you have these set up, do the following:

    git clone git://github.com/ql-io/ql.io.git
    cd ql.io
    make install

**Note** If you get "ERR! Error: EACCES, Permission denied" errors, please take a look 
at [npm issue #194](https://github.com/isaacs/npm/issues/194) and 
[what-no-sudo](http://foohack.com/2010/08/intro-to-npm/#what_no_sudo).

These steps will link ql.io modules locally so can you refer to those modules from your apps using
`npm link`.

To run tests

    make test

ql.io source is organized into several modules that you can test independently.

    cd modules/engine
    make test

## Using ql.io as a Stand-Alone Server

If you are interested in using ql.io as a stand-alone server, setup a new ql.io app and start the
server.

    mkdir myapp
    cd myapp
    curl https://raw.github.com/ql-io/ql.io/master/modules/template/init.sh | bash
    bin/start.sh

Using latest versions of Firefox or Chrome, go to
[http://localhost:3000](http://localhost:3000) to see ql.io's Web Console. See the
[Quickstart Guide](http://ql.io/docs/quickstart) for more details.</p>

## Using ql.io in a Node App

If you are interested in using ql.io in your node app, use

    npm install ql.io-engine

After that you can simply execute the core engine.
    
    var Engine = require('ql.io-engine');
    var engine = new Engine({
        connection: 'close'
    });

    var script = "create table geocoder " +
                 "  on select get from 'http://maps.googleapis.com/maps/api/geocode/json?address={address}&sensor=true' " +
                 "     resultset 'results.geometry.location'" +
                 "select lat as lattitude, lng as longitude from geocoder where address='Mt. Everest'";

    engine.execute(script, function(emitter) {
        emitter.on('end', function(err, res) {
            console.log(res.body[0]);
        });
    });

## Making Contributions

Fixes and features via pull requests are welcome as long as the contributor agrees to the
[Contributor License Agreement](https://github.com/downloads/ql-io/ql.io/ql.io-CLA.pdf). Print,
sign, and email a scanned copy to subbu/AT/ebaysf/DOT/com before submitting the first pull request.

To help move pull requests quickly, consider socializing your idea in the
[email group](http://groups.google.com/group/qlio).

## Discussions

Subscribe to the [google group](http://groups.google.com/group/qlio). 


CHANGES.md
## October 17, 2012

* now works with node-expat 2.0.0 

## August 27, 2012

* debugger and visualization in console.

## August 09, 2012

* Fix Issue #558

## August 09, 2012

* Node 0.8.x and other dependecy upgrades.
* Timeout support for update.

## JULY 23, 2012

*  Upgrade code-mirror to 2.32 to avoid space issue

## July 11, 2012

* Changes to mem-cache-local to queue operations if connections are not available.
* Upgrade code-mirror to 2.3

## July 5, 2012

* Update support.

## July 3, 2012

* Insert JSON feature.

## July 1, 2012

* Fix 493: ws events should include event name along with the payload

## June 29, 2012

* No Compression if the CPU load is > 50%
* Version 0.7

## June 25, 2012

* Retry once for idempotent requests on timeouts

## June 22, 2012

* Agent.maxSockets increased to 1000. See Issue: https://github.com/ql-io/ql.io/issues/512

## June 19, 2012

* Regression from previous fix. Remove duplicate click handlers.

## June 18, 2012

* Fix https://github.com/ql-io/ql.io/issues/13 - disable autorun. Also use `pushState` to store
  history.

## June 11, 2012

* Fix https://github.com/ql-io/ql.io/issues/478

## June 5, 2012

* Support C style block comments
* Add file path/name to comiple errors

## June 4, 2012

* Update versions
* Push fallback and optional params

## May 29, 2012

* Explicit dependencies between modules.

## May 25, 2012

* Update the context with the udf filtered data

## May 22, 2012

* Add fallback syntax to the language

## May 20, 2012

* Version 0.6

## May 18, 2012

* Support array style reference in columns clause, such as `select 'b-1', 'b-3'['c-1'] from a`.
* Disable ability to enable/disable ecv checks by default. You can turn it on by adding arg
   `--ecvControl true` to the start script.

## May 17, 2012

* Add optional parameters in route. Including "with optional params" in route would make params
  without ^ prefix optional.
* Add more test cases, some bug fix in app.js to fully support defaults.

## May 15, 2012

* Be able to start on multiple ports (separated by commas)
* Added support for multiple attachments.

## May 14, 2012

* Support larger than 1mb data in local-mem-cache.

## May 8, 2012

* Moved maxNestedRequests warning out of the _.each loop in select.js

## May 3, 2012

* Add support for multipart requests.

## Apr 27, 2012

* End pending connections on close after responses are written.
* Support cache events (hit, miss, new, error, info, heartbeat)

## Apr 26, 2012

* Switch to new cluster2

## Apr 25, 2012

* Added new syntaxes "with part" and opaque insert param.

## Apr 24, 2012

* Add a null check on join loop
* Fix expression parsing in string template so that a token like `"{obj.prop[?(@.price > 2)]}"` is
  valid

## Apr 20, 2012

* Add support for escaped quotes in string values

## Apr 19, 2012

* Updated PEG.js to 0.7. I see 50% improvement in speed which matters for string and uri templates
  as we don't yet cache parser output.
* Remove duplicates from in clause. 
* Use hasOwnProperty in place of prop lookup while joining

## Apr 18, 2012

* Deal with non UTF-8 encodings from upstream resources

## Apr 17, 2012

* When joining, use '==' to maintain backwards compat

## Apr 16, 2012

* Minor cleanup of event names and classes to make the API less specific.

## Apr 15, 2012

* Refactor logging to error, access, proxy and default logs. The proxy log file contains outgoing
  req/resp, access log contains incoming requests, error log contains all errors and warnings,
  and the rest go to ql.io.log. All these files are rotated.

## Apr 13, 2012

* Include a payload with begin events

## Apr 12, 2012

* Support local offset and limit

## Apr 11, 2012

* Cleanup logging (first phase)
* Fix the case of alias names with joins and UDFs.

## Apr 08, 2012

* Support null values in Values

## Apr 07, 2012

* Fix the issue posted in https://groups.google.com/forum/?fromgroups#!topic/qlio/JaMC1GVP1sA.
* Add UDFs in where clause to post process rows. You can either tweak or remove a row. See
  https://gist.github.com/2334012 for semantics of UDFs. UDF support for the where clause is
  still a TODO.

## Apr 02, 2012

* Experimental UDF support.

## Mar 30, 2012

* Release version 0.5.

## Mar 29, 2012

* Fix regression in uri-template formatter (https://github.com/ql-io/ql.io/issues/409).

## Mar 28, 2012

* Upgrade cluster2, and also pass ecvPath to cluster2
* Fix issue 404
* Cache config param for cache provider changed from cache.name to cache.impl
* If cache provider not found with the given token then try to find it relative to "node_modules" of process.cwd().

## Mar 25, 2012

* Add `using headers` clause on return statements. Here is an example.

    return 'hello' via route '/hello' using method get
        using headers 'Cache-Control' = 'max-age=3600'

## Mar 22, 2012

* Externally supply Cache to engine through config.
* Logging Enhancements. Lot of bugs fixed in hierarchical logging.

## Mar 21, 2012

* Fix https://github.com/ql-io/ql.io/issues/329.
* Compression support for ql.io clients. Responses for routes are compressed as per
  'accept-encoding' header. gzip and deflate are supported. Default is gzip.

## Mar 20, 2012

* Factor out cluster function into [cluster2](https://github.com/ql-io/cluster2). This change
  also moves ECV checks to cluster2.
* Caching support based on new `expires <seconds>` keyword ***create table*** (example below). `compute key` monkey patch to generate custom keys.

**create table**

create table auto.compute.key on select get from 'http://a.uri.net' ..other things .. `expires 10`;

**compute key** (monkey patch)

	exports['compute key'] = function(args) {
    	//return args.uri;
	    var key = [];
    	key.push(args.table);
	    key.push(args.uri);
    	key.push(JSON.stringify(args.params));
	    return(key.join(':'));
	};	

## Mar 19, 2012

* Remove extraneous event emitted while processing the where clause.

## Mar 18, 2012

* Fix https://github.com/ql-io/ql.io/issues/372.


modules/app/lib/main.js
/*
 * Copyright 2011 eBay Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var Cluster = require('cluster2'),
    fs = require('fs'),
    winston = require('winston'),
    os = require('os'),
    _ = require('underscore'),
    program = require('commander'),
    Console = require('ql.io-console'),
    assert = require('assert');

// Trap all uncaught exception here.
process.on('uncaughtException', function(error) {
    // TODO: Report to logger
    console.log('Uncaught error');
    console.log(error.stack || error);
});

exports.version = require('../package.json').version;

exports.exec = function() {
    var loggerFn, cb, opts;
    if(arguments.length === 1) {
        cb = arguments[0];
    }
    else if(arguments.length === 2) {
        cb = arguments[0];
        opts = arguments[1];
    }
    else if(arguments.length === 3) {
        loggerFn = arguments[0];
        cb = arguments[1];
        opts = arguments[2];
    }

    loggerFn = loggerFn || addFileLoggers;

    // Process command line args.
    var cwd = process.cwd();
    program.option('-C, --cluster', 'run in cluster').
        option('-c, --config <configFile>', 'path to config', cwd + '/../config/dev.json').
        option('-p, --port <port>', 'port to bind to', '3000').
        option('-m, --monPort <monPort>', 'port for monitoring', 3001).
        option('-t, --tables <tables>', 'path of dir containing tables', cwd + '/tables').
        option('-r, --routes <routes>', 'path of dir containing routes', cwd + '/routes').
        option('-x, --xformers <xformers>', 'path of dir containing xformers', cwd + '/config/xformers.json').
        option('-a, --ecvPath <ecvPath>', 'ecv path', '/ecv').
        option('-c, --ecvControl <ecvControl>', 'allow disabling ecv', false).
        option('-n, --noWorkers <noWorkers>', 'no of workers', os.cpus.length).
        option('-e, --disableConsole', 'disable the console', false).
        option('-q, --disableQ', 'disable /q', false);
    if(opts) {
        _.each(opts, function(opt) {
            program.option(opt[0], opt[1], opt[2], opt[3]);
        })
    }
    program.parse(process.argv);

    var ports = _.map(program.port.split(','), function(port) {
        return parseInt(port);
    });
    var options = {
        cluster: program.cluster,
        port: ports,
        monPort: parseInt(program.monPort),
        config: program.config,
        tables: program.tables,
        routes: program.routes,
        xformers: program.xformers,
        disableConsole: program.disableConsole,
        disableQ: program.disableQ,
        noWorkers: program.noWorkers,
        'request-id': program.requestId || 'Request-ID',
        loggerFn: loggerFn,
        ecv: {
            path: program.ecvPath,
            control: program.ecvControl,
            monitor: '/tables',
            validator: function(status, headers, data) {
                return JSON.parse(data);
            }
        },
        timeout: 300 * 1000 // Idle client socket timeout
    };
    options.__proto__ = program;

    var cluster = new Cluster(options);
    if(process.argv.indexOf('stop') >= 0) {
        cluster.stop(options);
    }
    else if(process.argv.indexOf('shutdown') >= 0) {
        cluster.shutdown(options);
    }
    else {
        var emitter;
        cluster.listen(
            // Create an app and call back
            function(cb2) {
                createConsole(options, cluster, function(app, monApp, e) {
                    emitter = e;
                    cb2(app, monApp);
                })
            },
            // Cluster is ready
            function(app) {
                if(cb) {
                    cb(app, program, emitter);
                }
            }
        );
    }
}

exports.addFileLoggers = addFileLoggers;
function addFileLoggers(options, emitter) {
    // Attach listeners for logging
    // Ensure logs dir.
    var logdir = false;
    try {
        fs.readdirSync(process.cwd() + '/logs');
        logdir = true;
    }
    catch(e) {
        try {
            fs.mkdirSync(process.cwd() + '/logs/', parseInt('755', 8));
            logdir = true;
        }
        catch(e) {
        }
    }
    var logger = createLogger(logdir, '/logs/ql.io.log');
    var accessLogger = createLogger(logdir, '/logs/access.log');
    var errLogger = createLogger(logdir, '/logs/error.log');
    var proxyLogger = createLogger(logdir, '/logs/proxy.log');

    logger.setLevels(winston.config.cli.levels);
    emitter.on('ql.io-begin-event', function (event, message) {
        if(_.isObject(message)) {
            message.eventId = event.eventId;
            message.pid = process.pid;
        }
        if(event.type === 'URL') {
            accessLogger.info(message)
        }
        else if(event.name === 'http-request') {
            proxyLogger.info(message)
        }
    });
    emitter.on('ql.io-end-event', function (event, message) {
        if(_.isObject(message)) {
            message.eventId = event.eventId;
            message.pid = process.pid;
            message.duration = event.duration;
        }
        if(event.type === 'URL') {
            accessLogger.info(message)
        }
        else if(event.name === 'http-request') {
            proxyLogger.info(message)
        }
    });

    emitter.on('ql.io-event', function (event, message) {
        logger.info(message || event);
    });

    emitter.on('info', function (event, message) {
        logger.info(message || event);
    });

    emitter.on('ql.io-error', function (event, message, err) {
        errLogger.info(message || event);
        if(err) {
            errLogger.error(err.stack || err);
        }
    });
    emitter.on('error', function (event, message) {
        errLogger.error(message || event);
    });

    emitter.on('fatal', function (event, message) {
        errLogger.error(message || event);
    });

    emitter.on('ql.io-warning', function (event, message) {
        var warn = errLogger.warn || errLogger.warning;
        warn(message || event);
    });
    emitter.on('warning', function (message) {
        var warn = errLogger.warn || errLogger.warning;
        warn(message);
    });
}

function createConsole(options, cluster, cb) {
    var disableConsole = Boolean(program.disableConsole);
    var disableQ = Boolean(program.disableQ);
    return new Console({
        loggerFn: function(emitter) {
            // Add loggers
            options.loggerFn.call(null, options, emitter);

            // Listen to cluster events
            cluster.on('died', function(pid) {
                emitter.emit('fatal', {
                    pid: pid,
                    message: 'Process died'
                });
            });
            cluster.on('forked', function(pid) {
                emitter.emit('info', {
                    pid: pid,
                    message: 'Worker forked'
                });
            });
            cluster.on('SIGTERM', function(pid) {
                emitter.emit('info', {
                    signal: 'SIGTERM',
                    pid: pid,
                    message: 'Shutting down'
                });
            });
            cluster.on('warning', function(message) {
                emitter.emit('warning', message);
            })
        },
        'tables': program.tables,
        'routes': program.routes,
        'config': program.config,
        'xformers': program.xformers,
        'enable console': !disableConsole,
        'enable q': !disableQ,
        'request-id': program.requestId,
        'log levels': require('winston').config.syslog.levels}, cb);
}

function createLogger(logdir, name) {
    var logger = logdir ? new (winston.Logger)({
        transports: [
            new (winston.transports.File)({
                filename: process.cwd() + name,
                maxsize: 1024000 * 5,
                colorize: false,
                json: true,
                timestamp: function () {
                    return new Date();
                }
            })
        ]
    }) : new (winston.Logger)();
    return logger;
}


